File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        dummy_node = ListNode()
        l3 = dummy_node
        while l1 or l2 or carry:
            current_sum = carry
            if l1:
                current_sum += l1.val
                l1 = l1.next
            
            if l2:
                current_sum += l2.val
                l2 = l2.next
                
            carry = current_sum // 10
            current_sum = current_sum % 10

            l3.next = ListNode(current_sum)
            l3 = l3.next
        
        result_node = dummy_node.next
        del dummy_node
        return result_node
            
             

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        int carry = 0;
        ListNode* dummy = new ListNode();
        ListNode* l3 = dummy;

        while (l1 || l2 || carry) {
            int sum = carry;
            if (l1) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2) {
                sum += l2->val;
                l2 = l2->next;
            }

            carry = sum / 10;
            sum = sum % 10;

            l3->next = new ListNode();
            l3->next->val = sum;
            l3->next->next = NULL;
            l3 = l3->next;
        }

        ListNode* result = dummy->next;
        delete dummy;
        return result;
    }
};